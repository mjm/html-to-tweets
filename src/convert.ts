import * as htmlparser from "htmlparser2"
import { URL } from "url"

export interface ConvertOptions {
  baseURL?: string
}

export function convert(
  html: string,
  options: ConvertOptions = {}
): ConvertResult {
  const handler = new TweetHandler(options)

  const parser = new htmlparser.Parser(handler as any, { decodeEntities: true })
  parser.write(html)
  parser.end()

  return handler.result
}

export interface ConvertResult {
  body: string
  mediaURLs: string[]
}

class TweetHandler {
  options: ConvertOptions
  text: string = ""
  mediaURLs: string[] = []
  urls: string[] = []
  linkedTwitterUser: string | null = null
  embeddedTweet: { url?: string } | null = null

  constructor(options: ConvertOptions) {
    this.options = options
  }

  onopentag(name: string, attrs: { [key: string]: string }) {
    switch (name) {
      case "a":
        if (attrs.href) {
          if (this.embeddedTweet) {
            if (!attrs.href.startsWith("https://twitter.com")) {
              return
            }

            this.embeddedTweet.url = attrs.href.replace(/(.*)\?.*$/, "$1")
            return
          }
          const twitterUser = getTwitterUsername(attrs.href)
          if (twitterUser) {
            this.linkedTwitterUser = twitterUser
          } else {
            this.addURL(attrs.href)
          }
        }
        break
      case "blockquote":
        if (attrs.class === "twitter-tweet") {
          this.embeddedTweet = {}
        } else {
          this.text += "“"
        }
        break
      case "img":
        if (attrs.src) {
          this.addMediaURL(attrs.src)
        }
        break
      case "iframe":
        if (attrs.src) {
          const youtubeURL = getYoutubeLink(attrs.src)
          if (youtubeURL) {
            this.addURL(youtubeURL)
          }
        }
        break
    }
  }

  ontext(text: string) {
    if (this.linkedTwitterUser || this.embeddedTweet) {
      return
    }

    // collapse all whitespace to a single space
    this.text += text.replace(/\s+/g, " ")
  }

  onclosetag(name: string) {
    switch (name) {
      case "p":
        this.text += "\n\n"
        break
      case "br":
        this.text += "\n"
        break
      case "a":
        if (this.linkedTwitterUser) {
          this.text += `@${this.linkedTwitterUser}`
          this.linkedTwitterUser = null
        }
        break
      case "blockquote":
        if (this.embeddedTweet) {
          if (this.embeddedTweet.url) {
            this.addURL(this.embeddedTweet.url)
          }
          this.embeddedTweet = null
        } else {
          this.text += "”"
        }
        break
    }
  }

  onend() {
    this.text = this.text.trim()
    this.text += ` ${this.urls.join(" ")}`
    this.text = this.text
      .split("\n")
      .map(l => l.trim())
      .join("\n")
  }

  get result(): ConvertResult {
    return {
      body: this.text,
      mediaURLs: this.mediaURLs,
    }
  }

  private addURL(url: string) {
    this.urls.push(this.fixURL(url))
  }

  private addMediaURL(url: string) {
    this.mediaURLs.push(this.fixURL(url))
  }

  private fixURL(url: string): string {
    return new URL(url, this.options.baseURL).toString()
  }
}

const twitterUserRegex = new RegExp(
  "^https?://(?:www.)?twitter.com/([A-Za-z0-9_]{1,15})/?$"
)

function getTwitterUsername(url: string): null | string {
  const match = url.match(twitterUserRegex)
  if (!match) {
    return null
  }

  return match[1]
}

const youtubeRegex = new RegExp(
  "^https://www.youtube.com/embed/([A-Za-z0-9_]+)"
)

function getYoutubeLink(url: string): null | string {
  const match = url.match(youtubeRegex)
  if (!match) {
    return null
  }

  const videoID = match[1]
  return `https://www.youtube.com/watch?v=${videoID}`
}
