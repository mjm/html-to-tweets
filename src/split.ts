import { parseTweet, ParsedTweet } from "twitter-text"
import { Tweet } from "./"
import { ConvertResult } from "./convert"

export function split(input: ConvertResult): Tweet[] {
  const [first, ...rest] = numberTweets(splitTweets(input.body))
  return [
    { action: "tweet", body: first, mediaURLs: input.mediaURLs },
    ...rest.map((t): Tweet => ({ action: "tweet", body: t, mediaURLs: [] })),
  ]
}

function splitTweets(text: string, tweets: string[] = []): string[] {
  const [first, rest] = splitOnce(text)
  tweets.push(first)

  if (rest) {
    return splitTweets(rest, tweets)
  }

  return tweets
}

function splitOnce(text: string): [string, string?] {
  const first = shorten(text)

  const second = text.substring(first.length).trim()
  if (second) {
    return [first, second]
  }

  return [first]
}

function shorten(text: string): string {
  let results = parseTweet(text)
  if (!results.valid) {
    // drop a word to make room for numbering
    text = dropWord(truncate(text, results))

    // keep dropping words until it fits
    while ((results = parseTweet(`${text} (XX/XX)`)) && !results.valid) {
      text = dropWord(text)
    }
  }

  return text
}

function truncate(
  text: string,
  { validRangeStart, validRangeEnd }: ParsedTweet
): string {
  return text.substring(validRangeStart, validRangeEnd)
}

function dropWord(text: string): string {
  return text.replace(/\s+\S*$/, "")
}

function numberTweets(tweets: string[]): string[] {
  const total = tweets.length

  if (total <= 1) {
    return tweets
  }

  return tweets.map((t, i) => `${t} (${i + 1}/${total})`)
}
