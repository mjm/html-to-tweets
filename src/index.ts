import { convert } from "./convert"
import { split } from "./split"

interface TranslateInput {
  title: string | null
  url: string
  html: string
}

export type TwitterAction = Tweet | Retweet

export interface Tweet {
  action: "tweet"
  body: string
  mediaURLs: string[]
}

export interface Retweet {
  action: "retweet"
  tweetID: string
}

export function translate(input: TranslateInput): TwitterAction[] {
  if (input.title) {
    return [
      {
        action: "tweet",
        body: `${input.title} ${input.url}`,
        mediaURLs: [],
      },
    ]
  }

  const result = convert(input.html, { baseURL: input.url })
  const tweets = split(result)
  if (tweets.length === 1) {
    const retweetID = extractRetweetId(tweets[0])
    if (retweetID) {
      return [
        {
          action: "retweet",
          tweetID: retweetID,
        },
      ]
    }
  }

  return tweets
}

const tweetRegex = new RegExp(
  "^https?://(?:www.)?twitter.com/[A-Za-z0-9_]+/status/([0-9]+)/?$"
)

function extractRetweetId(tweet: Tweet): string | null {
  if (tweet.mediaURLs.length) {
    // if the tweet has media, it's not a retweet
    return null
  }

  const match = tweet.body.match(tweetRegex)
  if (!match) {
    return null
  }

  return match[1]
}
