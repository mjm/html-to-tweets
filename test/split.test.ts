import { split } from "../src/split"

test("does not split if tweet is short enough", () => {
  const input = {
    body: "abcd ".repeat(56),
    mediaURLs: [],
  }

  expect(split(input)).toMatchSnapshot()
})

test("splits tweets along word boundaries", () => {
  const input = {
    body: "abcd ".repeat(57),
    mediaURLs: [],
  }

  expect(split(input)).toMatchSnapshot()
})

// Eventually this behavior should change, but this is it for now
test("attaches all media URLs to the first tweet", () => {
  const input = {
    body: "abcd ".repeat(57),
    mediaURLs: ["https://example.com/foo.jpg", "https://example.com/bar.jpg"],
  }

  expect(split(input)).toMatchSnapshot()
})

test("splits bacon ipsum", () => {
  const body =
    "Bacon ipsum dolor amet andouille rump tongue flank leberkas tail shoulder picanha cupim turducken hamburger brisket. Bacon pastrami capicola, pork chop venison landjaeger rump swine doner kevin frankfurter chuck strip steak jerky. Pork belly kielbasa pork buffalo bresaola. Tenderloin fatback short ribs meatloaf. Meatloaf sausage biltong bacon turkey cow frankfurter. Frankfurter jerky drumstick doner, bacon sausage turducken alcatra pig fatback strip steak."
  const input = { body, mediaURLs: [] }

  expect(split(input)).toMatchSnapshot()
})
