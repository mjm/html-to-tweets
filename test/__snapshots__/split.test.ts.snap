// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attaches all media URLs to the first tweet 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd (1/2)",
    "mediaURLs": Array [
      "https://example.com/foo.jpg",
      "https://example.com/bar.jpg",
    ],
  },
  Object {
    "action": "tweet",
    "body": "abcd abcd abcd (2/2)",
    "mediaURLs": Array [],
  },
]
`;

exports[`does not split if tweet is short enough 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd ",
    "mediaURLs": Array [],
  },
]
`;

exports[`splits bacon ipsum 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "Bacon ipsum dolor amet andouille rump tongue flank leberkas tail shoulder picanha cupim turducken hamburger brisket. Bacon pastrami capicola, pork chop venison landjaeger rump swine doner kevin frankfurter chuck strip steak jerky. Pork belly kielbasa pork buffalo (1/2)",
    "mediaURLs": Array [],
  },
  Object {
    "action": "tweet",
    "body": "bresaola. Tenderloin fatback short ribs meatloaf. Meatloaf sausage biltong bacon turkey cow frankfurter. Frankfurter jerky drumstick doner, bacon sausage turducken alcatra pig fatback strip steak. (2/2)",
    "mediaURLs": Array [],
  },
]
`;

exports[`splits tweets along word boundaries 1`] = `
Array [
  Object {
    "action": "tweet",
    "body": "abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd abcd (1/2)",
    "mediaURLs": Array [],
  },
  Object {
    "action": "tweet",
    "body": "abcd abcd abcd (2/2)",
    "mediaURLs": Array [],
  },
]
`;
